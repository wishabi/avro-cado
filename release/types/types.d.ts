/// <reference types="node" />
import * as Avro from "avsc";
export interface Options {
    schemaRegistry: string;
    numRetries: number;
    schema: any;
    subject: string;
    wrapUnions?: string;
}
export interface EncodeFunc {
    (message: object): Buffer;
}
export declare type ResolverMap = {
    [index: number]: Promise<Avro.Resolver>;
};
export interface DecoderInfo {
    subject: string;
    schema: Avro.Type;
    resolversMap: ResolverMap;
    createSchemaResolver: Function;
}
export interface DecodeFunc {
    (buffer: Buffer): Promise<Object>;
}
