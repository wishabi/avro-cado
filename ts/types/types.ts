import * as Avro from "avsc";

/*
 ********************************************************
 *                                          PACKAGE TYPES
 ********************************************************
 */

/*
 ********************************************************
 *                                          ENCODER TYPES
 ********************************************************
 */

export interface EncodeFunc {
  (message: object): Buffer;
}

/*
 ********************************************************
 *                                          DECODER TYPES
 ********************************************************
 */

export type ResolverMap = { [index: number]: Promise<Avro.Resolver> };

export interface DecoderInfo {
  subject: string;
  schema: Avro.Type | null;
  resolversMap: ResolverMap;
  retrieveSchema: Function;
  createSchemaResolver: Function;
  getSchemaResolver: Function;
}

export interface DecodeFunc {
  (buffer: Buffer): Promise<Object>;
}
