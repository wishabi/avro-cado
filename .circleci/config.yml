defaults: &defaults
  working_directory: ~/workspace
  docker:
      - image: circleci/node:10.14.2-alpine

version: 2
jobs:
  build:
    <<: *defaults
    steps:
      - checkout
      # see if we have the node_modules cached already
      - restore_cache:
          key: 'v1-node-module-cache-{{ .Branch }}-{{ checksum "package.json" }}'
      - run: npm i
      # save the node_modules cache
      - save_cache:
          key: 'v1-node-module-cache-{{ .Branch }}-{{ checksum "package.json" }}'
          paths:
            - ~/workspace/node_modules

      - persist_to_workspace:
          root: ~/workspace
          paths:
            - .
  test:
    <<: *defaults
    steps:
      - attach_workspace:
          at: ~/workspace
      - run:
          command: npm run test
      - store_test_results:
          path: reports/unit-tests
  lint:
    <<: *defaults
    steps:
      - attach_workspace:
          at: ~/workspace
      - run:
          command: npm run lint
workflows:
  version: 2
  build_and_test:
    jobs:
      - build
      - test:
          requires:
            - build