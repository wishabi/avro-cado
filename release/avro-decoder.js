"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Avro = require("avsc");
const rp = require("request-promise");
const config_1 = require("./config");
const util_1 = require("./util");
/**
 * Retrieve the Avro schema from the schema registry. On an error
 * eligible for retry, try up to the configure number of times
 *
 * @param id - the schema id that should be retrieved
 * @return - The Avro schema requested
 */
const genSchemaRetriever = ({ subject, schemaRegistry, numRetries }) => (id) => __awaiter(this, void 0, void 0, function* () {
    const req = {
        uri: `${schemaRegistry}/schemas/ids/${id}`,
        headers: [{ Accept: config_1.ACCEPT_HEADERS }],
        json: true,
        resolveWithFullResponse: true
    };
    let error = null;
    // implement retry on certain status/reason codes
    for (let i = 0; i <= numRetries; i += 1) {
        try {
            return JSON.parse((yield rp(req)).body.schema);
        }
        catch (err) {
            // save the error
            error = err;
            if (!util_1.handleError(err)) {
                break;
            }
        }
    }
    throw new Error(`Failed to retrieve schema for subject ${subject} with id ${id} :: ${error.message}`);
});
/**
 * Retrieve the Avro schema from the schema registry, parse it,
 * and create a resolver to the local Avro schema.
 *
 * @param subject - The subject under which the schema is registered
 * @param id - the schema id that should be the source for the
 *             schema resolver
 * @param schema - the target Avro schema for the resolver
 * @return - The Avro schema resolver used to convert messages
 *           encoded with the specified Avro schema to the local
 *           Avro schema format
 */
const genCreateSchemaResolver = ({ subject, retrieveSchema }) => (id, schema) => __awaiter(this, void 0, void 0, function* () {
    /**
     * Retrieve the schema by ID from the schema registry ...
     */
    const msgSchema = yield retrieveSchema(id);
    /**
     * ... parse the schema ...
     */
    const avSourceSchema = Avro.Type.forSchema(msgSchema);
    /**
     * ... and create a resolver to the schema we know how to consume
     */
    return schema.createResolver(avSourceSchema);
});
/**
 * Create and return an Avro schema resolver, if one doesn't already
 * exist for the schema id in the encoded message to the schema passed
 * in
 *
 * @param encoded - The Avro encoded buffer
 * @param schema - the target Avro schema for the resolver
 * @return - The promise that holds the Avro resolver
 */
const genGetSchemaResolver = ({ subject, resolversMap, createSchemaResolver }) => (encoded, schema) => __awaiter(this, void 0, void 0, function* () {
    // Do not attempt to get a resolver
    if (!schema) {
        return null;
    }
    if (!encoded) {
        return null;
    }
    /**
     * Extract the schema id with which the buffer was encoded
     */
    const id = encoded.slice(1, 5).readInt32BE(0);
    if (!resolversMap[id]) {
        /**
         * We do not have a resolver from the source schema to the
         * schema we support. Go and get a resolver.
         *
         * IMPORTANT
         *
         * We store a promise of a resolver in our hashmap. The reason is that
         * if we have multiple messages being handled asynchronously, we do not
         * want all of them to flood the schema registry with REST calls
         * to retrieve the schema
         */
        resolversMap[id] = createSchemaResolver(id, schema);
    }
    return resolversMap[id];
});
/**
 * Avro decode the passed in data
 *
 * @param encoded - The Avro encoded buffer
 * @param schema - the target Avro schema to which the decoded payload
 *                 should conform
 * @param resolver - a promise for an Avro resolver that will
 *                   convert the data from the schema with which
 *                   it was encoded to the schema which we want to
 *                   read
 * @return - The promise that holds the Avro decoded payload
 */
const decodePayload = (encoded, schema, resolver) => __awaiter(this, void 0, void 0, function* () {
    // Do not attempt to avro decode if we have no target schema specified
    if (!schema) {
        return encoded;
    }
    // Do not attempt to avro decode null value
    if (!encoded) {
        return encoded;
    }
    return schema.decode(encoded.slice(5), 0, yield resolver).value;
});
/**
 * Create an Avro decoder based on the specified parameters and
 * return a closure that takes an encoded Buffer and decodes it
 *
 * @param schema - the target Avro schema to which the decoded payload
 *                 should conform
 * @param getSchemaResolver - a function used to retrieve the resolver
 *                            object that will be used to decode the payload
 * @param buffer - The Avro encoded buffer
 * @return - The Avro decoded object that conforms to the specified schema
 */
const genMessageDecoder = ({ schema, getSchemaResolver }) => (buffer) => __awaiter(this, void 0, void 0, function* () {
    return yield decodePayload(buffer, schema, getSchemaResolver(buffer, schema));
});
/*****************************************************************/
/**                      EXPORTED INTERFACE                     **/
/*****************************************************************/
exports.createDecoder = (opts) => {
    // Aggregate the configuration values with defaults
    const mergedOpts = config_1.processOptions(opts);
    // decoder info
    const decoderInfo = {
        subject: mergedOpts.subject,
        schema: mergedOpts.schema ? Avro.Type.forSchema(mergedOpts.schema) : null,
        resolversMap: {},
        retrieveSchema: null,
        createSchemaResolver: null,
        getSchemaResolver: null
    };
    decoderInfo.retrieveSchema = genSchemaRetriever(mergedOpts);
    decoderInfo.createSchemaResolver = genCreateSchemaResolver(decoderInfo);
    decoderInfo.getSchemaResolver = genGetSchemaResolver(decoderInfo);
    return genMessageDecoder(decoderInfo);
};
//# sourceMappingURL=avro-decoder.js.map