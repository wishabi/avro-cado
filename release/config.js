"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("../ts/util");
exports.ACCEPT_HEADERS = "application/vnd.schemaregistry.v1+json, application/vnd.schemaregistry+json, application/json";
const optionsDefault = {
    schemaRegistry: "http://localhost:8081",
    numRetries: 10,
    wrapUnions: "auto",
    subject: null,
    schema: null
};
/**
 * Validate the options and merge them with the default values
 *
 * @param opts - the user specified options
 * @return - merged configuration options
 */
exports.processOptions = (opts) => {
    if (!opts.subject) {
        throw new Error("subject key missing");
    }
    if (!opts.schema) {
        throw new Error("schema key missing");
    }
    // Aggregate the configuration values with defaults
    return util_1.aggregateOptions(optionsDefault, opts);
};
//# sourceMappingURL=config.js.map