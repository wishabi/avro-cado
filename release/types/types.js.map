{"version":3,"file":"types.js","sourceRoot":"","sources":["../../ts/types/types.ts"],"names":[],"mappings":""}