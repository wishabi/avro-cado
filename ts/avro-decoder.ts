import * as Avro from "avsc";
import * as rp from "request-promise";
import { ACCEPT_HEADERS, processOptions } from "./config";
import {
  DecodeFunc,
  DecoderInfo,
  Options,
  SchemaResolverFunc
} from "./types/types";
import { aggregateOptions, handleError } from "./util";

/**
 * Retrieve the Avro schema from the schema registry. On an error
 * eligible for retry, try up to the configure number of times
 *
 * @param subject - the subject under which the schema is registered
 * @param schemaRegistry - the schema registry path
 * @param numRetries - the number of retries
 * @param id - the schema id that should be retrieved
 * @return - The Avro schema requested
 */
export const retrieveSchema = async (
  { subject, schemaRegistry, numRetries }: Options,
  id: number
) => {
  const req = {
    uri: `${schemaRegistry}/schemas/ids/${id}`,
    headers: { Accept: ACCEPT_HEADERS },
    json: true,
    resolveWithFullResponse: true
  };

  let error = null;

  // implement retry on certain status/reason codes
  for (let i = 0; i <= numRetries; i += 1) {
    try {
      return JSON.parse((await rp(req)).body.schema);
    } catch (err) {
      // save the error
      error = err;

      if (!handleError(err)) {
        break;
      }
    }
  }

  throw new Error(
    `Failed to retrieve schema for subject ${subject} with id ${id} :: ${
      error.message
    }`
  );
};

/**
 * Retrieve the Avro schema from the schema registry, parse it,
 * and create a resolver to the local Avro schema.
 *
 * @param id - the schema id that should be the source for the
 *             schema resolver
 * @param schema - the target Avro schema for the resolver
 * @return - The Avro schema resolver used to convert messages
 *           encoded with the specified Avro schema to the local
 *           Avro schema format
 */
export const genCreateSchemaResolver = (
  opts: Options
): SchemaResolverFunc => async (
  id: number,
  schema: Avro.Type
): Promise<Avro.Resolver> => {
  /**
   * Retrieve the schema by ID from the schema registry ...
   */
  const msgSchema = await retrieveSchema(opts, id);

  /**
   * ... parse the schema ...
   */
  try {
    const avSourceSchema = Avro.Type.forSchema(msgSchema);

    /**
     * ... and create a resolver to the schema we know how to consume
     */
    return schema.createResolver(avSourceSchema);
  } catch (err) {
    throw new Error(
      `Local schema is not compatible with schema from registry with id ${id} :: ${
        err.message
      }`
    );
  }
};

/**
 * Create an Avro decoder based on the specified parameters and
 * return a closure that takes an encoded Buffer and decodes it
 *
 * @param schema - the target Avro schema to which the decoded payload
 *                 should conform
 * @param createSchemaResolver - a function used to create the resolver
 *                               object that will be used to decode the payload
 * @param resolversMap - a hashmap keys on the schema ids where the value is a
 *                       promise holding a schema resolver
 * @param buffer - The Avro encoded buffer
 * @return - The Avro decoded object that conforms to the specified schema
 */
export const genPayloadDecoder = ({
  schema,
  createSchemaResolver,
  resolversMap
}: DecoderInfo): DecodeFunc => async (
  buffer: Buffer | null
): Promise<object> => {
  // Do not attempt to avro decode null value
  if (!buffer) {
    return buffer;
  }

  /**
   * Extract the schema id with which the buffer was encoded
   */
  const id: number = buffer.slice(1, 5).readInt32BE(0);

  if (!resolversMap[id]) {
    /**
     * We do not have a resolver from the source schema to the
     * schema we support. Go and get a resolver.
     *
     * IMPORTANT
     *
     * We store a promise of a resolver in our hashmap. The reason is that
     * if we have multiple messages being handled asynchronously, we do not
     * want all of them to flood the schema registry with REST calls
     * to retrieve the schema
     */
    resolversMap[id] = createSchemaResolver(id, schema);
  }

  return schema.decode(buffer.slice(5), 0, await resolversMap[id]).value;
};

/*
 *****************************************************************
 *                      EXPORTED INTERFACE                       *
 *****************************************************************
 */

export const createDecoder = (opts: Options): DecodeFunc => {
  // Aggregate the configuration values with defaults
  const mergedOpts: Options = processOptions(opts);

  // decoder info
  const decoderInfo: DecoderInfo = {
    subject: mergedOpts.subject,
    schema: Avro.Type.forSchema(mergedOpts.schema),
    resolversMap: {},
    createSchemaResolver: null
  };

  decoderInfo.createSchemaResolver = genCreateSchemaResolver(mergedOpts);

  return genPayloadDecoder(decoderInfo);
};
