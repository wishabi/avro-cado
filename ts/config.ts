import { aggregateOptions } from "../ts/util";
import { IOptions } from "./types/types";

export const ACCEPT_HEADERS: string =
  "application/vnd.schemaregistry.v1+json, application/vnd.schemaregistry+json, application/json";

const optionsDefault: IOptions = {
  schemaRegistry: "http://localhost:8081",
  numRetries: 10, // number of attempts to call schemaRegistry
  wrapUnions: "auto", // avsc option
  subject: null, // subject for schema registration
  schema: null // schema object as needed for avsc
};

/**
 * Validate the options and merge them with the default values
 *
 * @param opts - the user specified options
 * @return - merged configuration options
 */
export const processOptions = (opts: IOptions): IOptions => {
  if (!opts.subject) {
    throw new Error("subject key missing");
  }

  if (!opts.schema) {
    throw new Error("schema key missing");
  }

  // Aggregate the configuration values with defaults
  return aggregateOptions(optionsDefault, opts);
};
