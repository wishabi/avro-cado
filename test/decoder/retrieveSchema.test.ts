jest.mock("axios");

import axios from "axios";
import { retrieveSchema } from "../../ts/avro-decoder";
import { ACCEPT_HEADERS } from "../../ts/config";
import { Options } from "../../ts/types/types";
import {
  RETRY_ERROR_CODE_50002,
  RETRY_ERROR_CODE_50003,
  RETRY_STATUS_CODE_500
} from "../../ts/util";

const opts: Options = {
  subject: "subject",
  schemaRegistry: "host",
  numRetries: 1,
  schema: { data: "schema" }
};

const SCHEMA_ID = 1;

describe("genSchemaRetriever", () => {
  beforeEach(() => {
    axios.mockReset();
  });

  it("should get the schema on success", async () => {
    axios.mockImplementationOnce((params) => {
      return {
        data: {
          schema: JSON.stringify(opts.schema)
        }
      };
    });

    expect.assertions(3);

    const schema = await retrieveSchema(opts, SCHEMA_ID);

    expect(axios).toHaveBeenCalledTimes(1);

    expect(schema).toMatchSnapshot();
    expect(axios).toHaveBeenCalledWith({
      headers: { Accept: ACCEPT_HEADERS },
      json: true,
      method: "get",
      url: `${opts.schemaRegistry}/schemas/ids/${SCHEMA_ID}`
    });
  });

  /*
   * Retry once and on a retriable error and then fail
   * one all configured retries have been exhausted
   */
  it("should retry 1 time on retriable error code 50002", async () => {
    axios
      .mockImplementationOnce(() => {
        throw {
          response: {
            status: RETRY_STATUS_CODE_500,
            data: {
              error_code: RETRY_ERROR_CODE_50002,
              message: "500:50002"
            }
          },
          message: "500:50002"
        };
      })
      .mockImplementationOnce(() => {
        throw {
          response: {
            status: RETRY_STATUS_CODE_500,
            data: {
              error_code: RETRY_ERROR_CODE_50002,
              message: "500:50002"
            }
          },
          message: "500:50002"
        };
      });

    expect.assertions(2);

    await expect(retrieveSchema(opts, SCHEMA_ID)).rejects.toMatchSnapshot();

    expect(axios).toHaveBeenCalledTimes(2);
  });

  it("should retry 1 time on retriable error code 50003", async () => {
    axios
      .mockImplementationOnce((params) => {
        throw {
          response: {
            status: RETRY_STATUS_CODE_500,
            data: {
              error_code: RETRY_ERROR_CODE_50003,
              message: "500:50003"
            }
          },
          message: "500:50003"
        };
      })
      .mockImplementationOnce((params) => {
        throw {
          response: {
            status: RETRY_STATUS_CODE_500,
            data: {
              error_code: RETRY_ERROR_CODE_50003,
              message: "500:50003"
            }
          },
          message: "500:50003"
        };
      });

    expect.assertions(2);

    await expect(retrieveSchema(opts, SCHEMA_ID)).rejects.toMatchSnapshot();

    expect(axios).toHaveBeenCalledTimes(2);
  });

  /*
   * Do NOT retry on a NON retriable error even though
   * retries are configured
   */
  it("should retry 0 times on fatal error", async () => {
    axios.mockImplementationOnce((params) => {
      throw {
        message: "600:60002",
        response: {
          status: 600,
          data: {
            error_code: RETRY_ERROR_CODE_50002,
            message: "600:60002"
          }
        }
      };
    });

    expect.assertions(2);

    await expect(retrieveSchema(opts, SCHEMA_ID)).rejects.toMatchSnapshot();

    expect(axios).toHaveBeenCalledTimes(1);
  });
});
