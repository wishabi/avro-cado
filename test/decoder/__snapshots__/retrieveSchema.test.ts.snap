// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`genSchemaRetriever should get the schema on success 1`] = `
Object {
  "data": "schema",
}
`;

exports[`genSchemaRetriever should retry 0 times on fatal error 1`] = `[Error: Failed to retrieve schema for subject subject with id 1 :: 600:50002]`;

exports[`genSchemaRetriever should retry 1 time on retriable error code 50002 1`] = `[Error: Failed to retrieve schema for subject subject with id 1 :: 500:50002]`;

exports[`genSchemaRetriever should retry 1 time on retriable error code 50003 1`] = `[Error: Failed to retrieve schema for subject subject with id 1 :: 500:50003]`;
